libs/web3/src/lib/use-bridge-contract.ts

```diff
@@ -39,3 +52,165 @@ export const useBridgeContract = (allowDefaultProvider = false) => {
 
   return contract;
 };
+
+type CollateralBridgeConfig = {
+  chainId: number;
+  address: string;
+  confirmations: number;
+};
+
+export const useCollateralBridgeConfigs = () => {
+  const {
+    config: ethConfig,
+    loading: ethLoading,
+    error: ethError,
+  } = useEthereumConfig();
+  const {
+    configs: evmConfigs,
+    loading: evmLoading,
+    error: evmError,
+  } = useEVMBridgeConfigs();
+
+  const configs = useMemo(() => {
+    const configs: CollateralBridgeConfig[] = [];
+    if (ethConfig) {
+      configs.push({
+        chainId: Number(ethConfig.chain_id),
+        address: ethConfig.collateral_bridge_contract.address,
+        confirmations: ethConfig.confirmations,
+      });
+    }
+    if (evmConfigs) {
+      for (const evm of evmConfigs) {
+        configs.push({
+          chainId: Number(evm.chain_id),
+          address: evm.collateral_bridge_contract.address,
+          confirmations: evm.confirmations,
+        });
+      }
+    }
+    return compact(uniq(configs));
+  }, [ethConfig, evmConfigs]);
+
+  return {
+    configs,
+    loading: ethLoading || evmLoading,
+    error: ethError || evmError,
+  };
+};
+
+export const useCollateralBridge = (chainId?: number) => {
+  const { provider, chainId: activeChainId } = useWeb3React();
+  const { configs, error: configsError } = useCollateralBridgeConfigs();
+
+  const config = useMemo(() => {
+    if (configsError || !chainId) return undefined;
+    return configs.find((c) => c.chainId === chainId);
+  }, [chainId, configs, configsError]);
+
+  const signer = useMemo(() => {
+    return provider?.getSigner();
+  }, [provider]);
+
+  const contract = useMemo(() => {
+    if (!provider || !config || chainId !== activeChainId) return undefined;
+    return new CollateralBridge(config.address, signer || provider);
+  }, [activeChainId, chainId, config, provider, signer]);
+
+  return {
+    contract,
+    chainId,
+    address: config?.address,
+    config,
+    error: chainId && chainId !== activeChainId ? ERR_WRONG_CHAIN : undefined,
+  };
+};
+
+export const useProvider = (chainId?: number, allowDefaultProvider = false) => {
+  const { provider: activeProvider, chainId: activeChainId } = useWeb3React();
+  const { providers: defaultProviders } = useDefaultWeb3Providers();
+
+  const [error, setError] = useState<Error | undefined>(undefined);
+
+  const defaultProvider = useMemo(() => {
+    if (chainId && defaultProviders) {
+      return defaultProviders[chainId];
+    }
+    return undefined;
+  }, [chainId, defaultProviders]);
+
+  const provider = useMemo(() => {
+    if (chainId && (activeProvider || defaultProvider)) {
+      if (chainId === activeChainId) {
+        return activeProvider;
+      } else if (allowDefaultProvider) {
+        return defaultProvider;
+      }
+    }
+    return undefined;
+  }, [
+    activeChainId,
+    activeProvider,
+    allowDefaultProvider,
+    chainId,
+    defaultProvider,
+  ]);
+
+  useEffect(() => {
```

Comment:
Non blocking: Might save some render cycles if you just derive an error?

```
  let error
  if (chainId && !isSupportedChainId(chainId)) {
    error = ERR_CHAIN_NOT_SUPPORTED
  }
  else if (activeChainId !== chainId) {
    error = ERR_WRONG_CHAIN
  }

```

Reply:
removed that code, not needed anymore